webroot/results/admin/persons_finish_unfinished.php

```diff
@@ -128,24 +128,58 @@ function showUnfinishedPersons () {
   $caseNr = 0;
+  $availableSpots = array(); // array of semiIds in progress
   foreach( $personsFromResultsWithoutId as $person ){
     list( $name, $countryId, $firstYear ) = $person;
     
     #--- Try to compute the semi-id.
-    $quarterId = removeUglyAccentsAndStuff( extractRomanName( $name ));
-    $quarterId = preg_replace( '/[^a-zA-Z ]/', '', $quarterId );
-    $quarterId = strtoupper( substr( preg_replace( '/(.*)\s(.*)/', '$2$1', $quarterId ), 0, 4 ));
-    if ( strlen ( $quarterId ) == 0 ) {
-      // if the name comes empty, invent a quarterId
-      $quarterId = 'XXXX';
-    } else if ( strlen( $quarterId ) < 4 ) {
-      // make sure the quarterId is 4-letter long
-      while ( strlen( $quarterId ) < 4 ) {
-        $quarterId .= $quarterId;
-      }
-      $quarterId = substr( $quarterId, 0, 4 );
+    $paddingLetter = 'U';
+    $neatName = strtoupper(preg_replace('/[^a-zA-Z ]/','',removeUglyAccentsAndStuff(extractRomanName($name))));
+    $nameParts = explode(' ',$neatName);
+    $lastName = $nameParts[count($nameParts)-1];
+    $restOfName = '';
+    for ($i=0;$i<count($nameParts)-1;$i++) {
+        $restOfName .= $nameParts[$i];
+    }
+    // follows a simple trick that prevents us from empty or too short restOfNames and provides the appropriate padding
+    $restOfName .= str_repeat($paddingLetter,4);
+    $lettersToShift = max(0,4-strlen($lastName));
+    $cleared = false;
+    while (!$cleared && $lettersToShift<=4) {
+        $quarterId = substr($lastName,0,4-$lettersToShift) . substr($restOfName,0,$lettersToShift);
+        $semiId = $firstYear . $quarterId;
+        // has the semiId been produced before from persons in progress?
+        if (array_key_exists($semiId,$availableSpots)) {
+            if ($availableSpots[$semiId]) {
+                $availableSpots[$semiId]--;
+                $cleared = true;
+            } else {
+                $lettersToShift++;
+            }
+        } else {
+            // else search the db
+            $lastIdTaken = dbQuery("SELECT id FROM Persons WHERE id LIKE '".$semiId."__' ORDER BY id DESC LIMIT 1");
+            if (!count($lastIdTaken)) {
+                $cleared = true;
+                $availableSpots[$semiId] = 98;
+            } else {
+                $counter = intval(substr($lastIdTaken[0]['id'],8,2),10);
+                if ($counter==99) {
+                    $availableSpots[$semiId] = 0;
+                    $lettersToShift++;
+                } else {
+                    $availableSpots[$semiId] = 98-$counter;
+                    $cleared = true;
+                }
+            }
+        }
+    }
+    if (!$cleared) {
+        // if we didn't clear a spot (almost impossible) we stick with the first combination
+        $lettersToShift = max(0,4-strlen($lastName));
+        $semiId = $firstYear . substr($lastName,0,4-$lettersToShift) . substr($restOfName,0,$lettersToShift);
+        $availableSpots[$semiId] = 0;
```

Comment:
In fact we know *for sure* that it conflicts.  It's just that the script has tried all the possibilities and none of them was valid.  If we reach here with `$cleared` set to false (something that is not going to happen in centuries) then the person posting will have the same problem that Baiqiang had recently, and the software team of the future will have work to do.